Refactor to support multiple instances + general improvements

JamesWrigley opened the pull request:
The main purpose of this PR was to refactor the vagrant module to properly support multiple instances with multi-machine `Vagrantfile`'s, though it ended up changing a lot of other things too (sorry about that).

Overview of changes:
- Use templating exclusively instead of the vagrant.yml config file, which has
  been removed. This reduces the amount of code duplication, and makes it simpler to debug bad `Vagrantfile`'s.
- Multiple instances are supported using multi-machine `Vagrantfile`'s instead of overwriting the config files for each instance. This should fix any problems with parallel VM creation.
- Made most of the module options per-instance options to support multiple
  instances better.
- Renamed the `interfaces` to `networks` and changed the structure to:
   ```
   networks:
     - identifier: <network identifier>
       options:
         <option name>: <option value>
   ```
  This matches the naming in the Vagrant [network documentation](https://www.vagrantup.com/docs/networking/basic_usage.html).
- Support for [vagrant-cachier](https://github.com/fgrehm/vagrant-cachier) was removed since the project is unmaintained.
- Enabling VirtualBox's `linked_clone` option by default was removed to simplify the code, since the user can still enable it if they want (not sure about this, maybe we should enable it by default for convenience?).
- The provider setting is now enforced by passing it to `Vagrant.up()`.
- Removed `VagrantClient._status()` since `Vagrant.status()` returns what we want.
- Moved to Python 3 by default.
- Add support for disabling parallel VM creation.
- Re-enabled all of the tests (de6f180), which should now pass.

apatard commented:
I've not yet looked at the code but I've did some testing:
- took a role I have with very basic multi instance molecule test: it runs fine with this
- executed the testsuite (lint, packaging, py37, devel, distros) and found some issues.

The CI issues are:
  * the CI is using only libvirt for now, so default and multi-nodes scenario are failing for this reason. I've tried locally to add vbox to run all tests with libvirt ***and*** vbox but the build of the vbox kernel modules was a quite complicated so I gave up. 
  * The ``config_options`` scenario is failing as the instance configuration in the ``molecule.yml`` with your patch is not compatible with current ``molecule.yml`` files. Excerpt of generated ``Vagrantfile``:
```yaml
    c.vm.synced_folder ".", "/vagrant", disabled: true

    # Config options
    c.synced_folder = true
```
  It's due to:

```yaml
    config_options:
      synced_folder: true
```
  * the ``provider_config_option`` failure is, imho, a variation around the option case. The generated file looks like:
```yaml
      # General provider options
      provider.nic_model_type = ""rtl8139""
```
It's due to:
```yaml
    provider_options:
      nic_model_type: '"rtl8139"'
```
In this case, vagrant status returns a syntax error message leading to:
```shell
    The error was: AttributeError: 'NoneType' object has no attribute 'splitlines'
```
python-vagrant is not handling the syntax error returned by vagrant properly ?

With the multinode test disabled and the other ``molecule.yml`` fixed, the CI looks happy.

So, do we really want to break exising setups by changing the ``molecule.yml`` file ? We're still in 0.X, which means it's fine from a semver point of view. If everybody is fine with this breakage, I won't object but please at least fix the CI ``molecule.yml`` files.
Extra question: In the multinode test, I see ``virtualbox__intnet`` is it something specific to vbox ? it's only a name, right ?

I'm going to look at the code soon.

apatard commented on molecule_vagrant/playbooks/create.yml:
```diff
@@ -25,7 +25,7 @@
 
         provision: "{{ item.provision | default(omit) }}"
-
+        parallel: "{{ molecule_yml.driver.provider.parallel | default(yes) }}"
```
nitpick : maybe ``default(omit)`` ?

JamesWrigley replied:
Fixed in 79db427. Did the same for `provider.name`.

apatard commented on molecule_vagrant/modules/vagrant.py:
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/python3
```
What about putting python3 changes in a different commit ?

JamesWrigley replied:
Fixed in 72aae24.

apatard commented on molecule_vagrant/modules/vagrant.py:
```diff
@@ -147,216 +145,72 @@
+
+    # Config options
+    {% for key, value in instance.config_options.items() -%}
+    c.{{ key }} = {{ ruby_format(value) }}
```
I guess it's also related to the global comment I gave about configuration format change and I need to check it won't prevent some configuration case (like ``config.ssh`` ?). Shouldn't it be `` c.vm.{{ key }}`` ?

JamesWrigley replied:
I don't think so, otherwise users wouldn't be able to access all the other config namespaces (`config.ssh`, `config.vagrant`, etc). The `config_options` variable expects the options to include the namespace, e.g. `ssh.remote_user: vagrant` or `vm.hostname: foo`.

apatard commented on molecule_vagrant/modules/vagrant.py:
```diff
@@ -147,216 +145,72 @@
+    {{ 'c.vm.box_url = "{}"'.format(instance.box_url) if instance.box_url }}
+
+    c.vm.hostname = "{{ instance.name }}"
+    c.vm.synced_folder ".", "/vagrant", disabled: {{ ruby_format(not instance.synced_folder) }}
```
If someone wants to change the synced folder, how can it be done ?

JamesWrigley replied:
They would have to set it in `instance_raw_config_args`, e.g.:
```yml
instance_raw_config_args:
  - 'vm.synced_folder "foo/", "/bar/"
```

apatard commented on molecule_vagrant/modules/vagrant.py:
```diff
@@ -508,79 +351,57 @@ def _get_config(self):
+            if provider.startswith("vmware_")
+            else "memory"
+        )
+
```
too bad that vmware needs a special case. I hope that vmware is the only one. At least with your change, instead of having to modify the ``Vagrantfile`` template for every provider, we only have to deal with exceptions, so that's nice.

JamesWrigley replied:
Full disclosure, I haven't tested this with VMWare. This logic was in the original code so I simply carried it over here.

apatard commented on molecule_vagrant/modules/vagrant.py:
```diff
@@ -508,79 +351,57 @@ def _get_config(self):
-
     def _get_vagrant(self):
-        v = vagrant.Vagrant(
+        return vagrant.Vagrant(
```
Shouldn't this kind of change go to the cleanup patch ?

JamesWrigley replied:
Fixed in 69ba9bd.

apatard commented on molecule_vagrant/modules/vagrant.py:
```diff
@@ -508,79 +351,57 @@ def _get_config(self):
+        return [
+            {
+                "name": instance["name"],
+                "box": instance.get("box", "generic/alpine310"),
```
hm. I'm not convinced that hard coding the default image in the python module is a good idea.
If the image gets removed:
- if it's located in the ``create.yml``, it's easy for a user to override, while waiting for us to fix.
- if it's in the python module, it's harder imho.

At least, I understand that it should prevent issues like bug #19

JamesWrigley replied:
The reason it's hardcoded in the module is because it can no longer be specified directly in `create.yml`, since it's part of the new `instances` variable instead. However, perhaps we should have an explicit check in the module for required fields in the `instances` variable? I think it makes sense to require at least `box` to be set by the user.

apatard commented on molecule_vagrant/modules/vagrant.py:
```diff
@@ -589,48 +410,33 @@ def _get_stderr_log(self):
-            provider_override_args=dict(type="list", default=None),
-            provider_raw_config_args=dict(type="list", default=None),
+            instances=dict(type="list", required=True),
+            provider=dict(type="str", default="virtualbox"),
```
In a perfect world the parameters should have remained identical but I think it's not possible at all.
So, maybe, the only thing to do is to document somewhere that things have changed between 0.3 and 0.4 to avoid bugs from people with their own ``create.yml`` / ``destroy.yml`` playbooks.

JamesWrigley replied:
Added a note about it in 79db427.

apatard commented on molecule_vagrant/modules/vagrant.py:
```diff
@@ -52,77 +52,40 @@
-  - Supports check mode. Run with --check and --diff to view config difference,
-    and list of actions to be taken.
+  # - Supports check mode. Run with --check and --diff to view config difference,
+  #   and list of actions to be taken.
```
afair, check mode and diff are not supported, what about just removing theses lines ?

JamesWrigley replied:
Removed in 6e20017.

ssbarnea commented:
recheck

JamesWrigley commented:
> the CI is using only libvirt for now, so default and multi-nodes scenario are failing for this reason. I've tried locally to add vbox to run all tests with libvirt _**and**_ vbox but the build of the vbox kernel modules was a quite complicated so I gave up.

Yeah, didn't realize that. Switched them to use libvirt in 2fee04d.

> The `config_options` scenario is failing as the instance configuration in the `molecule.yml` with your patch is not compatible with current `molecule.yml` files.

Fixed in 2fee04d.

> the `provider_config_option` failure is, imho, a variation around the option case. The generated file looks like:

Fixed in 2fee04d, I believe the issue was that the option value was double-quoted, which caused a syntax error in the templated Vagrantfile.

> So, do we really want to break exising setups by changing the `molecule.yml` file ? We're still in 0.X, which means it's fine from a semver point of view. If everybody is fine with this breakage, I won't object but please at least fix the CI `molecule.yml` files.

IMHO it is acceptable to break the `platforms` option since it lets us handle multi-machine Vagrantfiles properly. I am mostly in favour of breaking the `interfaces` option since it fits better with the Vagrant documentation, but I don't feel very strongly about that.

> Extra question: In the multinode test, I see `virtualbox__intnet` is it something specific to vbox ? it's only a name, right ?

Yep, it's a [VBox specific thing](https://www.vagrantup.com/docs/providers/virtualbox/networking#virtualbox-internal-network). Removed in 2fee04d.

---

One note about f6af150 and 2fee04d, I found that vagrant-libvirt (1.2) is quite flaky on my machine when it comes to assigning custom IP addresses to VMs, especially when creating/destroying in parallel. So I made the `parallel` option global, and set `parallel: no` in the `multi-node` scenario, which made it much more reliable on my machine. We'll see how the CI likes it :)

I also found some weirdness when getting synced folders to work, Vagrant was using NFS v4 for some reason which is apparently incompatible with my system (Fedora 32), so I set a default of `nfs_version: 3` in e12d979. This matches the [Vagrant docs](https://www.vagrantup.com/docs/synced-folders/nfs#nfs_version) so I think it's a reasonable setting.

apatard commented:
recheck

apatard commented on molecule_vagrant/test/scenarios/molecule/multi-node/molecule.yml:
```diff
@@ -1,41 +1,48 @@
-    config_options:
-      synced_folder: true
+    provider_options:
+      qemu_use_session: false
```
This ``qemu_use_session`` thing is missing on config_options scenario too, leading to test failure. In fact, I think that it possibly needs to be handled in the template. After all, it's a mix of possibly upstream bug and of a change in fedora packaging:
https://fedoraproject.org/wiki/Changes/Vagrant_2.2_with_QEMU_Session
https://github.com/vagrant-libvirt/vagrant-libvirt/issues/958
With this quick and dirty change, it seems to be fine:
```diff
--- a/molecule_vagrant/modules/vagrant.py
+++ b/molecule_vagrant/modules/vagrant.py
@@ -159,6 +159,12 @@ Vagrant.configure('2') do |config|
       {% for option, value in instance.provider_options.items() -%}
       provider.{{ option }} = {{ ruby_format(value) }}
       {% endfor %}
+      # https://fedoraproject.org/wiki/Changes/Vagrant_2.2_with_QEMU_Session
+      # https://bugzilla.redhat.com/show_bug.cgi?id=1697773
+      # https://github.com/vagrant-libvirt/vagrant-libvirt/issues/958
+      {% if provider == "libvirt" and 'qemu_use_session' not in instance.provider_options %}
+      provider.qemu_use_session = false
+      {% endif %}
 
       # Raw provider args
       {% for arg in instance.provider_raw_config_args -%}
```

apatard replied:
more thoughts about this: I'm wondering if a fix for this issue needs to be pushed in a different PR, and then you'll have to modify your PR. After all, this has nothing to do with your work. What do you think ? If you agree with me, I'll make the PR.

ssbarnea commented:
I am closing this PR because it was not updated in a long time and because in its current form it does not pass requirements for being accepted: it combines totally unrelated changes.

Please have a read on https://medium.com/@hugooodias/the-anatomy-of-a-perfect-pull-request-567382bb6067 or other similar articles, the key part is "Single Responsibility Principle". 

Please feel free to open new PRs that are small and easy to review.
